### ISSUE TITLE
I have computed histograms and confusion matrix on validation dataset

### ISSUE THREAD

**Issue by hcl14** [idealo/image-quality-assessment#18]
I am attempting to train model on AVA myself and faced very low quality of predictions. Digging further, I found predictions to be very strange and started to investigate pretrained models and asking for help.

Let me introduce my research on the example of your model which has very similar performance to mine. 

At first, let's take a look on a histogram of predicted mean scores and standard deviations (Fig.9 of original paper):

![original](https://user-images.githubusercontent.com/2828322/50369047-4b6b4580-0598-11e9-8da5-5355b0458f13.png)

Here is your one, built on validation dataset you provided:

![screenshot_20181222_031920](https://user-images.githubusercontent.com/2828322/50369067-840b1f00-0598-11e9-8701-b61f9ea4eb40.png)

Mine looks similar as well. From the histogram one can decide that the model does not output scores > 7 and <3, but otherwise it should mirror real distribution well.

However, when I started to check images for my model manually, I found that scores seem very scattered and look often totally inadequate.

Let's compute accuracy of predicted scores as `Mean(1. - |m' - m|/m)`, where `m'` is predicted score and `m` is ground truth.  For your model it is 91.1%, for my model it is 90.8%. Seems good so far, but we need to remember that for score 5 it means mean error 5 - 0.9*5 = 0.5, and is even bigger for tails of the dataset distribution. Standard deviation of differences between scores is 0.37.

But let's compute confusion matrix (mine is similar to yours):
```

Got labels: 25548
[[   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0   18  100   84    9    0    0    0    0    0]
 [   1   21  518 3027 1832  178    2    0    0    0]
 [   0    7  144 3472 9527 2660  129    0    0    0]
 [   0    0    3  147 1683 1786  198    2    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]]
accuracy:0.9111294033307482
standard deviation of score differences:0.3676722622236499
```
And to compare with, let's **randomly shuffle predictions**:
```
random shuffle:
[[   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    5   53  109   40    4    0    0    0]
 [   0    9  188 1440 2874  992   76    0    0    0]
 [   1   31  460 4201 8144 2903  197    2    0    0]
 [   0    6  112 1036 1924  689   52    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]
 [   0    0    0    0    0    0    0    0    0    0]]
accuracy:0.8577966426127535
standard deviation of score differences:0.5579391415989003
```

We can see that score 5 still dominates because of it's abundance in dataset. Others are not much different from model prediction. Accuracy is 0.86. Also, I want to mention that when I tried to create balanced validation set with sets of scores <4, 4-7 and >7 to be represented equally, I've got 83% accuracy and 0.5 std of differences which is equal to the result above. 

I've shared all the code on gist:
https://gist.github.com/hcl14/d641f82922ce11cee0164b16e6786dfb

Also here are correlation coefficients for scores predictions:
```
Pearson: (0.6129669517144579, 0.0)
Spearman: SpearmanrResult(correlation=0.5949598193491837, pvalue=0.0)
```
The article also reports values 0.5-0.6

Would be great to hear some insights on that. Is it overfit cause by Adam optimizer and we really needed to optimize it via SGD with lr=10-7 as in the parer?

**Comment by hcl14** [comment]
HI @clennan, thanks for your answer. The reason why I did random shuffle of predictions is that I was confused by the fact that histogram resembles target distribution of scores very closely (I have the same for my MobileNet v2 model), but correlation is just 0.6, like in Google paper. 

I did strong oversampling of tails (elements with scores <4 and >7), but that did not change much for me, just model starts to assign low and high scores, but with the same corr=0.6.

I have provided rank Spearman correlation for your model SRCC=0.59, which is close to Google. In my model after long training with manual tweaks of learning rate and batch size I have SRCC=0.64 on the same validation dataset as yours.

**Comment by hcl14** [comment]
Also, as long as we use pretrained ImageNet model, I used subtraction of Imagenet Mean and division on ImageNet variance (values can be googled) for images, instead of just bringing them to `[-1,1]` as vanilla `.preprocess_inputs()` function does.

**Comment by clennan** [comment]
Cool that you managed to improve the performance, is the SRCC=0.64 with MobileNet or MobileNet v2? Would you be interested in contributing the model to the repository? :)

I think the oversampling of tail samples provided already an improvement, even though the rank correlations do not improve much, but absolute values do, which makes the classifier more useful in practice (eg when working with thresholds). 

If you want to improve the model further my suggestion is to focus on a specific domain (eg hotels, buildings, or people) and collect domain specific training data. It becomes a much easier classification problem then.

And `.preprocess_input()` does not only scale to `[-1, 1]`, but also does the mean substraction for colour channels.

**Comment by clennan** [comment]
Ah, interesting, I didn't know that MobileNetV2 uses Inception preprocessing, instead of VGG preprocessing (which normalizes with ImageNet parameters) like MobileNet (https://github.com/keras-team/keras-applications/blob/df0e26c73951e107a84969944305f492a9abe6d7/keras_applications/imagenet_utils.py#L157).

Let me know when you decided on the model version you would like to publish :)

**Comment by hcl14** [comment]
Hi @clennan , here I tried to put all my thoughts on AVA and TID2013 together and provided pretrained models:

https://github.com/hcl14/AVA-and-TID2013-image-quality-assessment

I refer to your repository frequently, as our models work very similar, on my opinion.

**Comment by clennan** [comment]
Hi, a couple of comments on your analysis for the aesthetic model. Discussing performance is always difficult as expectations differ so much, so this is just my opinion :)

- if a rank correlation coefficient of 0.6 is garbage for you, then that’s understandable, but I disagree with your statement that it is inevitable to get garbage results from such a model. The aesthetic model tries to predict something very abstract and subjective, and is being trained on a very diverse dataset (AVA). To me this seems like an impossible task, way harder than any image recognition, so my expectations were low to begin with. And what you consider garbage was actually not garbage for us. We tested some hotel images on the model trained solely on AVA and were surprised how good it ranked them, see here e.g. bedrooms

![image](https://user-images.githubusercontent.com/19587525/50630462-45367b00-0f41-11e9-8f26-8ca1a3f125be.png)

- as with any ML model the most important performance driver is the data you feed it - so once we narrowed down the domain (hotel images) and put together our own dataset, the model’s performance improved and we were seeing rank correlations around 0.77. So we took inspiration from a weak performing general aesthetic model and made it provide business value for a specific domain. My advice to you is don’t spend so much time optimising learning rates etc. but find an application domain for your ML model (e.g. hotels, cars) and label your own dataset.

**Comment by clennan** [comment]
Interesting idea - maybe it would make sense to take the image embedding straight from the im2txt model and add it to the image embedding from the Nima model, and then use dense layers to predict aesthetic scores?! The im2txt embedding, that is used to generate the descriptions, might include information that helps to classify aesthetics, similar to your approach with word vectors, but might be the more direct way to incorporate this information into Nima

**Comment by hcl14** [comment]
Yes, I thought about network embedding, but 

1) what I understood from the way how im2txt was trained, significant part of the performance can be already gained from training LSTM over frozen network, with weights from ImageNet (i.e. features are unchanged),
2) LSTM in im2txt uses predefined vocabulary (COCO dataset), assigning those words to image features (without word2vec or any extra knowledge).

My idea is that we can dig into human associations for those pictures using another knowledge domain, which is relevant - word2vec context vectors. Let's suppose that some objects like "piece of paper", "letters" or "humans with flowers instead of their heads" (example images from original paper) in image description both refer to something not aesthetic (writing) or controversial (unusual combination of objects) and drag human opinion score down / increase variance - then this dependency may be theoretically captured by pretrained word embeddings, like GoogleNews and more or less generalized to other objects because of cosine similarity of embeddings for contextually related words.

Of course, maybe you are right, and special embedding can be trained somehow, but I don't have good ideas at the moment.
